### PR TITLE
Polyfil test heirarchies

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
This allows extension authors to only need to provide a workspace `TestHeirachy` and when opening a document it will create a default document `TestHeirarchy` from the workspace one thus "polyfilling" the document hierarchies.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
 			"restart": true,
 			"name": "Attach to Extension Host",
 			"timeout": 30000,
-			"port": 5870,
+			"port": 5555,
```
Comment (TylerLeonhardt) [inline .vscode/launch.json:20]: will revert

File: src/vs/workbench/api/common/extHostTesting.ts [hunk 2]
```diff
@@ -222,6 +228,110 @@ export class ExtHostTesting implements ExtHostTestingShape {
...
+			this.hasNodeMatchingFilter = !this.parent
+				|| this.matchesUri
+				|| this.getWrappedChildren().some(child => child.doesHaveNodeMatchingFilter());
+		}
+
+		return this.hasNodeMatchingFilter;
+	}
+
+	constructor(public readonly actual: vscode.TestItem, private filterToUri: vscode.Uri, private readonly parent?: TestItemFilteredWrapper) {
+		this.matchesUri = this.actual.location?.uri.toString() === this.filterToUri.toString();
```
Comment (connor4312) [inline src/vs/workbench/api/common/extHostTesting.ts:328]: Though unlikely, it's possible the test location will change and start matching the filter, so this should probably be evaluated at check time.

File: src/vs/workbench/api/common/extHostTesting.ts [hunk 3]
```diff
@@ -222,6 +228,110 @@ export class ExtHostTesting implements ExtHostTestingShape {
...
+			return;
+		}
+
+		const onDidChangeTest = new Emitter<vscode.TestItem>();
+		workspaceHierarchy.onDidChangeTest(node => {
+			const wrapper = TestItemFilteredWrapper.getWrapperForTestItem(node, document.uri);
+			if (wrapper.doesHaveNodeMatchingFilter(true)) {
+				onDidChangeTest.fire(wrapper);
+			} else {
+				onDidChangeTest.fire(wrapper.visibleParent);
```
Comment (connor4312) [inline src/vs/workbench/api/common/extHostTesting.ts:248]: Related to the comment down below -- it's possible that the location will change for an existing testing. If this happens:

- If the before and after of `doesHaveNodeMatchingFilter` is the same, do nothing, otherwise;
- If the test was filtered out before, we should invalidated the cached `hasNodeMatchingFilter` up to the former `visibleParent` before firing the event (so the differ will see a new child that needs to be published);
- If the test existed before, do the same thing, but up to the _current_ visibleParent (so the subtree will be hidden)

**Review by connor4312 — COMMENTED** [review]
Looks nice overall, just one edge case :)

File: src/vs/workbench/api/common/extHostTesting.ts (hunk 3 above)
Comment (TylerLeonhardt) [reply src/vs/workbench/api/common/extHostTesting.ts:258]: this should be addressed now

File: src/vs/workbench/api/common/extHostTesting.ts [hunk 4]
```diff
@@ -222,6 +228,159 @@ export class ExtHostTesting implements ExtHostTestingShape {
...
+
+	// Reset the cache of whether or not you can see a node from a particular node
+	// up to it's visible parent.
+	public reset(): void {
+		if (this === this.visibleParent) {
+			this.matchesFilter = undefined;
+		} else {
+			this.parent?.reset();
+			this.matchesFilter = undefined;
+		}
```
Comment (connor4312) [inline src/vs/workbench/api/common/extHostTesting.ts:374]: nit: can be simplified slightly

```suggestion
		if (this !== this.visibleParent) {
			this.parent?.reset();
		}
		this.matchesFilter = undefined;
```

**Review by connor4312 — APPROVED** [review]
nice work!

File: src/vs/workbench/api/common/extHostTesting.ts [hunk 5]
```diff
@@ -222,6 +228,159 @@ export class ExtHostTesting implements ExtHostTestingShape {
...
+		};
+	}
+}
+
+/*
+ * A class which wraps a vscode.TestItem that provides the ability to filter a TestItem's children
+ * to only the children that are located in a certain vscode.Uri.
+ */
+class TestItemFilteredWrapper implements vscode.TestItem {
+	private static wrapperMap = new WeakMap<vscode.Uri, WeakMap<vscode.TestItem, TestItemFilteredWrapper>>();
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/workbench/api/common/extHostTesting.ts]: done

File: src/vs/workbench/api/common/extHostTesting.ts (hunk 4 above)
Comment (TylerLeonhardt) [reply src/vs/workbench/api/common/extHostTesting.ts]: fixed

**Review by connor4312 — APPROVED** [review]
